Help usage and shellcheck fixes

jasonkarns opened the pull request:
shellcheck fixes

fixed some things that [shellcheck](http://www.shellcheck.net/) static analysis found
- SC2164: Use cd ... || exit in case cd fails.
- SC2035: Use ./\* so names with dashes won't become options.
- SC2086: Double quote to prevent globbing and word splitting.
- SC2166: Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.
- SC2155: Declare and assign separately to avoid masking return values.
- SC2035: Use ./_._._-_ so names with dashes won't become options.
- SC2035: Use ./_-_._._ so names with dashes won't become options.

only print --usage help, not entire help output when given unknown args

remove extraneous 'exec'

Tests are failing on master and failing for identical reasons on this branch.

jasonkarns committed:
usage help

unknown args:
- trigger help to stderr, not stdout
- only print --usage, not full help
- exit as error

add explicit --help option
- prints to stdout, not stderr
- prints full help, not just usage
- clean exit

jasonkarns committed:
shell string manipulation instead of sed

substring removal via string manipulation instead of sed

tpope commented:
The `--usage` addition makes sense, and the replacement of `sed` is a nice touch. I'm not sure about the rest.

* `exec` means one less process running. Does it hurt anything?
* `*` to `./*` doesn't matter because `echo_lines` doesn't accept options. Plus it muddies up the echo with `./`
* I know of portability problems with `-a`, but I don't think they apply when running under bash.
* The extra quotes don't hurt anything, but I don't see what they help, given that the referenced variables are guaranteed not to contain spaces best I can tell.

jasonkarns commented:
Most the changes (except `exec`), were caught with the `shellcheck` tool.

- I removed `exec` because it seemed inconsistent with other invocations. The only rationale I have against it is that it makes the exit implicit. Indeed, this PR itself is broken because I forgot to add `exit`s when I removed `exec` :/ . And the exit status is carried from the exec'd command; which may or may not be desired. I'm assuming it's not desired in this case, because it's only emitting help usage. (An error while emitting help usage probably shouldn't cascade?) I'll update the PR per your direction, but I would suggest it be a bit more consistent with `exec` usage (using it in more places where early exit is desired).

- `echo_lines` doesn't accept options, but that doesn't mean that whatever it may be piped to (in the future) does not. `echo_lines_without_symlinks` doesn't accept options, either, but its output is filtered (unmodified) and immediately piped to `sed`, which does take options.

- I'm generally against `-a` for stylistic reasons in addition to portability concerns, but that's subjective, naturally. Up to you.

- Again, my personal preference is to excessively double quote since in the process of maintenance and future changes, what was once unnecessary can quickly become necessary (and buggy). The quotes don't hurt, yet help prevent future bugs in such a case as they become necessary.

tpope commented:
* If help is invoked due to invalid arguments, the exit status *should* be a failure, actually. So we should drop the `exec` and add an `exit 1` afterwards (which means expanding the `||` case to a proper `if`).
* `sed` doesn't read arguments from stdin so I don't understand your concern about piping here.
* I don't think stylistic preference is reason enough to change existing, functioning code in either of these cases.

jasonkarns commented:
> If help is invoked due to invalid arguments, the exit status should be a failure, actually. So we should drop the exec and add an exit 1 afterwards (which means expanding the || case to a proper if).

good call

> sed doesn't read arguments from stdin so I don't understand your concern about piping here.

yep, brain fart

> I don't think stylistic preference is reason enough to change existing, functioning code in either of these cases.

legit

Changes forthcoming

jasonkarns committed:
unalias help

- add explicit --help option (stdout, full help, clean exit)
- help usage for unknown args (stderr, just usage, error exit)
- update help to come from unalias script and not alias script

tpope commented:
Please clean up the history. One commit would be fine or you can try to split it into a small number.

jasonkarns commented:
Do you want the fail-helper refactor included or no?

tpope commented:
Uh I guess leave it out for now.

jasonkarns commented:
squashed to 3 commits
